### ISSUE TITLE
Type checking should be done on `__post_init__` signature for `dataclass` classes

### ISSUE THREAD

**Issue by bryanforbes** [microsoft/pyright#2360]
**Describe the solution you'd like**
Given the following code:

```python
from __future__ import annotations

from collections.abc import Iterable
from dataclasses import InitVar, dataclass, field


@dataclass
class A:
    items: list[int]

    def __post_init__(self, items: list[int]) -> None:
        self.items = list(items)


@dataclass
class B:
    iterable: InitVar[Iterable[int]]

    items: list[int] = field(init=False)

    def __post_init__(self) -> None:
        self.items = []


@dataclass
class C:
    iterable: InitVar[Iterable[int]]

    def __post_init__(self, iterable: Iterable[str]) -> None:
        self._items = list(iterable)


a = A([1, 2, 3, 4]) # TypeError: __post_init__() missing 1 required positional argument: 'items'
b = B((1, 2, 3, 4)) # TypeError: __post_init__() takes 1 positional argument but 2 were given
c = C((1, 2, 3, 4))
c._items  # list[str]
```

Currently, no type checking is done on the signature of `__post_init__` and no errors are generated for the above code. However, the instantiations for `A` and `B` will raise a `TypeError` because the signatures do not match what `dataclass` expects, and the type of `c._items` is incorrect because the type of the `iterable` parameter in `C` does not match the `InitVar` declaration. The type checker should produce an error for all three of these scenarios.

**Comment by erictraut** [comment]
This is a pretty esoteric feature of dataclasses. I've rarely seen it used. Mypy doesn't have any support for checking this either, and I don't think anyone has ever requested it (based on a search of their issue database). Furthermore, this isn't a false positive issue.

Without further signal from other pyright or mypy users, I don't think we could justify adding this functionality. We can revisit this in the future if and when we receive more requests or upvotes to this issue.

**Comment by erictraut** [comment]
This is included in pyright 1.1.201, which I just published. It will also be included in the next release of pylance.
